packages/flutter/test/widgets/overlay_portal_test.dart

```diff
@@ -1365,6 +1365,200 @@ void main() {
+                      controller: controller1,
+                      overlayChildBuilder: (BuildContext context) => const Placeholder(),
+                    ),
+              ),
+            ],
+          ),
+        ),
+      ),
+    );
+
+    expect(find.byType(Placeholder), findsOneWidget);
+    expect(controller1.isShowing, equals(true));
+  });
+
+  testWidgets('attachTarget is restored after reparenting', (WidgetTester tester) async {
+    final GlobalKey<State<OverlayPortal>> portalKey = GlobalKey<State<OverlayPortal>>();
+    final RenderBox childBox = RenderConstrainedBox(additionalConstraints: const BoxConstraints());
+    final RenderBox overlayChildBox = RenderConstrainedBox(
+      additionalConstraints: const BoxConstraints(),
+    );
+
+    late StateSetter setState;
+    bool moveToSecondOverlay = false;
+
+    final Widget child = WidgetToRenderBoxAdapter(renderBox: childBox);
+    final Widget overlayChild = WidgetToRenderBoxAdapter(renderBox: overlayChildBox);
+
+    late OverlayEntry overlayEntry1, overlayEntry2;
+    addTearDown(() {
+      overlayEntry1.dispose();
+      return overlayEntry2.dispose();
+    });
+
+    await tester.pumpWidget(
+      Directionality(
+        textDirection: TextDirection.ltr,
+        child: StatefulBuilder(
+          builder: (BuildContext context, StateSetter stateSetter) {
+            setState = stateSetter;
+            return Stack(
+              children: <Widget>[
+                Overlay(
+                  initialEntries: <OverlayEntry>[
+                    overlayEntry1 = OverlayEntry(
+                      builder: (BuildContext context) {
+                        return !moveToSecondOverlay
+                            ? OverlayPortal(
+                              key: portalKey,
+                              controller: controller1,
+                              overlayChildBuilder: (BuildContext context) => overlayChild,
+                              child: child,
+                            )
+                            : const SizedBox();
+                      },
+                    ),
+                  ],
+                ),
+                Overlay(
+                  initialEntries: <OverlayEntry>[
+                    overlayEntry2 = OverlayEntry(
+                      builder: (BuildContext context) {
+                        return moveToSecondOverlay
+                            ? OverlayPortal(
+                              key: portalKey,
+                              controller: controller1,
+                              overlayChildBuilder: (BuildContext context) => overlayChild,
+                              child: child,
+                            )
+                            : const SizedBox();
+                      },
+                    ),
+                  ],
+                ),
+              ],
+            );
+          },
+        ),
+      ),
+    );
+
+    // Move to second overlay
+    setState(() {
```

Comment:
I suspect when the `StatefulBuilder` rebuilds, the builder callback will be called a second time, creating new OverlayEntries so in the test 4 OverlayEntries were created but only 2 were disposed.

Reply:
Yaa, I need to give GlobalKey to `Overlay` widgets such that they also maintain themself until test is completed.

Reply:
Even after adding global key, tests are failing

Reply:
Global keys don't prevent rebuilds. Can you try creating the OverlayEntries at the start of the test, (and make them final), to make sure only two OverlayEntries are created in the test?

Reply:
Sure, and yaa even i was doing it wrongly with global keys. But yaa let me do the change.